spring-cloud-gcp-pubsub-stream-binder/src/test/java/org/springframework/cloud/gcp/stream/binder/pubsub/PubSubEmulator.java

```diff
@@ -0,0 +1,216 @@
+			Process psProcess = new ProcessBuilder("ps", "-v").start();
+			new BufferedReader(new InputStreamReader(psProcess.getInputStream()))
+					.lines()
+					.filter(psLine -> psLine.contains(hostPortParams))
+					.map(psLine -> new StringTokenizer(psLine).nextToken())
+					.forEach(pid -> {
+						this.killProcess(pid);
+					});
+		}
+		catch (IOException e) {
+			LOGGER.error("Failed to cleanup: ", e);
+		}
+		this.emulatorProcess.destroy();
+	}
+
+	/**
+	 * Returns an already-configured emulator binder when called from within a JUnit method.
+	 */
+	public PubSubTestBinder getBinder() {
+		return this.binder;
+	}
+
+	/**
+	 * Waits until a PubSub emulator configuration file is present.
+	 *
+	 * Fails if the file does not appear after 10 seconds.
+	 *
+	 * @throws InterruptedException which should interrupt the peaceful slumber and bubble up
+	 * to fail the test.
+	 */
+	private void waitForConfigCreation() throws InterruptedException {
+		int attempts = 10;
+		while (!Files.exists(this.EMULATOR_CONFIG_PATH) && --attempts >= 0) {
+			Thread.sleep(1000);
+		}
+		if (attempts < 0) {
+			throw new RuntimeException(
+					"Emulator could not be configured due to missing env.yaml. Are PubSub and beta tools installed?");
+		}
+	}
+
+	/**
+	 * Waits until a PubSub emulator configuration file is updated.
+	 *
+	 * Fails if the file does not update after 1 second.
+	 *
+	 * @throws InterruptedException which should interrupt the peaceful slumber and bubble up
+	 * to fail the test.
+	 */
+	private void waitForConfigUpdate(WatchService watchService) throws InterruptedException {
+		int attempts = 10;
+		while (--attempts >= 0) {
+			WatchKey key = watchService.poll(100, TimeUnit.MILLISECONDS);
+
+			if (key != null) {
+				Optional<Path> configFilePath = key.pollEvents().stream().filter(
+						event -> event.kind() == StandardWatchEventKinds.ENTRY_MODIFY)
+						.map(event -> (Path) event.context())
+						.filter(path -> ENV_FILE_NAME.equals(path.toString()))
+						.findAny();
+				if (configFilePath.isPresent()) {
+					return;
+				}
+			}
+		}
+
+		if (attempts < 0) {
+			throw new RuntimeException("Configuration file update could not be detected");
+		}
+	}
+
+	/**
+	 * Remembers host/port combination for future process cleanup.
+	 *
+	 * Validates that localhost is the only valid value; this will stop the test if emulator
+	 * decides to start with an IPv6 ::1 address.
+	 *
+	 * @param emulatorHostPort typically "localhost:8085"
+	 */
+	private void extractTeardownParams(String emulatorHostPort) {
+		int portSeparatorIndex = emulatorHostPort.indexOf(":");
+		if (portSeparatorIndex < 0 || !emulatorHostPort.contains("localhost")) {
+			throw new RuntimeException("Malformed host, can't create emulator: " + emulatorHostPort);
+		}
+		this.emulatorHost = emulatorHostPort.substring(0, portSeparatorIndex);
+		this.emulatorPort = emulatorHostPort.substring(portSeparatorIndex + 1);
```

Comment:
Why can't we just use `split(":")` here and check for the size of the result array?